pkg/flow/internal/controller/node_config_import.go

```diff
@@ -0,0 +1,328 @@
+		}
+	}
+}
+
+// processDeclareBlock stores the declare definition in the importedDeclares.
+func (cn *ImportConfigNode) processDeclareBlock(stmt *ast.BlockStmt) {
+	if _, ok := cn.importedDeclares[stmt.Label]; ok {
+		level.Error(cn.logger).Log("msg", "declare block redefined", "name", stmt.Label)
+		return
+	}
+	cn.importedDeclares[stmt.Label] = stmt.Body
+}
+
+// processDeclareBlock creates an ImportConfigNode child from the provided import block.
+func (cn *ImportConfigNode) processImportBlock(stmt *ast.BlockStmt, fullName string) {
+	sourceType := importsource.GetSourceType(fullName)
+	if _, ok := cn.importConfigNodesChildren[stmt.Label]; ok {
+		level.Error(cn.logger).Log("msg", "import block redefined", "name", stmt.Label)
+		return
+	}
+	childGlobals := cn.globals
+	// Children have a special OnBlockNodeUpdate function which notifies the parent when its content changes.
+	childGlobals.OnBlockNodeUpdate = cn.onChildrenContentUpdate
+	cn.importConfigNodesChildren[stmt.Label] = NewImportConfigNode(stmt, childGlobals, sourceType)
+}
+
+// evaluateChildren evaluates the import nodes managed by this import node.
+func (cn *ImportConfigNode) evaluateChildren() error {
+	for _, child := range cn.importConfigNodesChildren {
+		err := child.Evaluate(&vm.Scope{
+			Parent:    nil,
+			Variables: make(map[string]interface{}),
+		})
+		if err != nil {
+			return fmt.Errorf("imported node %s failed to evaluate, %v", child.label, err)
+		}
+	}
+	return nil
+}
+
+// onChildrenContentUpdate notifies the parent that the content has been updated.
+func (cn *ImportConfigNode) onChildrenContentUpdate(child BlockNode) {
+	// If the node is already updating its content, it will call OnBlockNodeUpdate
+	// so the notification can be ignored.
+	if !cn.inContentUpdate.Load() {
+		cn.OnBlockNodeUpdate(cn)
+	}
+}
+
+// Run runs the managed source and the import children until ctx is
+// canceled. Evaluate must have been called at least once without returning an
+// error before calling Run.
+//
+// Run will immediately return ErrUnevaluated if Evaluate has never been called
+// successfully. Otherwise, Run will return nil.
+func (cn *ImportConfigNode) Run(ctx context.Context) error {
+	cn.mut.Lock()
+	importChildrenCount := len(cn.importConfigNodesChildren)
+	cn.mut.Unlock()
+	if cn.source == nil {
+		return ErrUnevaluated
+	}
+
+	newCtx, cancel := context.WithCancel(ctx)
+	defer cancel() // This will stop the children and the managed source.
+
+	errChan := make(chan error, 1)
+
+	if importChildrenCount > 0 {
+		go func() {
+			errChan <- cn.runChildren(newCtx)
+		}()
+	}
+
+	go func() {
+		errChan <- cn.source.Run(newCtx)
+	}()
+
+	err := <-errChan
+
+	if err != nil {
+		level.Error(cn.logger).Log("msg", "import exited with error", "err", err)
+	} else {
+		level.Info(cn.logger).Log("msg", "import exited")
+	}
+	return err
+}
+
+// runChildren run the import nodes managed by this import node.
```

Comment:
Is it possible to replace the implementation here to use [our runner package](https://github.com/grafana/agent/blob/main/pkg/runner/runner.go)? I've noticed a few bugs in the code, but rather than point them out and check the rest of the code, I think it makes sense to align on the existing concept we have for managing a dynamic set of things to run.